File: haskell/p-56.hs
{-# LANGUAGE UnicodeSyntax #-}

import Prelude.Unicode

data Tree a = Empty | Branch a (Tree a) (Tree a)
            deriving (Show, Eq)


mirror ∷ Tree t → Tree t → Bool
mirror Empty Empty = True
mirror Empty _ = False
mirror _ Empty = False
mirror (Branch _ l1 r1) (Branch _ l2 r2) =
    mirror r1 l2 ∧ mirror l1 r2

symmetric ∷ Tree t → Bool
symmetric Empty = True
symmetric (Branch _ left right) = mirror left right

File: prolog/p-56.pl
mirror(nil, nil) :- !.
mirror(t(_, nil, nil), t(_, nil, nil)) :- !.
mirror(t(_, L1, R1), t(_, L2, R2)) :-
	mirror(R1, L2),
	mirror(L1, R2).

symmetric(nil) :- !.
symmetric(t(_, L, R)) :-
	mirror(L, R).

